File: assets/js/calculators.js
(function () {
  function ingredientMap() {
    return (window.DD_DATA && window.DD_DATA.ingredientMap) || {};
  }

  function effectiveCapacity(recipe, settings) {
    const regular = Number(recipe.glassCapacity) || Number(settings.glassCapacity) || 300;
    return recipe.ice === "full" ? (Number(settings.icedLiquidCapacity) || 150) : regular;
  }

  function estimateAmounts(recipe, settings) {
    const capacity = effectiveCapacity(recipe, settings || {});
    let total = 0;
    const amounts = (recipe.ingredients || []).map(item => {
      let volume = 0;
      if (item.unit === "ml") volume = Math.max(0, Number(item.amount) || 0);
      else if (item.unit === "fill") volume = Math.max(0, capacity * Math.max(0, Math.min(1, Number(item.fillTo) || 0)) - total);
      else if (item.unit === "top") volume = Math.max(0, capacity - total);
      total += volume;
      return Object.assign({}, item, { estimatedMl: volume });
    });
    return { capacity, total, amounts, exceedsCapacity: total > capacity + 0.01 };
  }

  function estimateAbv(recipe, settings, pantry) {
    const result = estimateAmounts(recipe, settings || {});
    const map = ingredientMap();
    let pureAlcohol = 0;
    result.amounts.forEach(item => {
      const stored = pantry && pantry[item.id] ? pantry[item.id] : {};
      const catalogAbv = map[item.id] ? Number(map[item.id].abv) || 0 : 0;
      const abv = stored.abv !== undefined && stored.abv !== "" ? Number(stored.abv) || 0 : catalogAbv;
      pureAlcohol += item.estimatedMl * abv / 100;
    });
    return Object.assign({}, result, { abv: result.total > 0 ? pureAlcohol / result.total * 100 : 0, pureAlcohol });
  }

  function estimateCost(recipe, settings, pantry) {
    const result = estimateAmounts(recipe, settings || {});
    let cost = 0;
    let pricedVolume = 0;
    let neededVolume = 0;
    result.amounts.forEach(item => {
      if (!item.estimatedMl) return;
      neededVolume += item.estimatedMl;
      const entry = pantry && pantry[item.id] ? pantry[item.id] : {};
      const price = Number(entry.price);
      const packSize = Number(entry.packSize);
      if (price >= 0 && packSize > 0 && entry.price !== "") {
        cost += price / packSize * item.estimatedMl;
        pricedVolume += item.estimatedMl;
      }
    });
    return Object.assign({}, result, { cost, coverage: neededVolume > 0 ? pricedVolume / neededVolume : 0, hasCost: pricedVolume > 0 });
  }

  function isOwned(pantry, id) {
    const entry = pantry && pantry[id];
    if (!entry || !entry.owned) return false;
    if (entry.stock === "" || entry.stock === undefined || entry.stock === null) return true;
    return Number(entry.stock) > 0;
  }

  function requiredIngredientIds(recipe) {
    return [...new Set((recipe.ingredients || [])
      .filter(item => !item.optional && item.id !== "ice")
      .map(item => item.id))];
  }

  function matchRecipe(recipe, pantry) {
    const required = requiredIngredientIds(recipe);
    const missing = required.filter(id => !isOwned(pantry || {}, id));
    const matched = required.length - missing.length;
    return {
      required,
      missing,
      matched,
      ratio: required.length ? matched / required.length : 1,
      status: missing.length === 0 ? "makeable" : (missing.length === 1 ? "almost" : "missing")
    };
  }

  function normalizeSearch(value) {
    return String(value || "")
      .normalize("NFD")
      .replace(/[\u0300-\u036f]/g, "")
      .toLowerCase()
      .replace(/[’'·()（）&+]/g, " ")
      .replace(/\s+/g, " ")
      .trim();
  }

  function searchableText(recipe, language) {
    const map = ingredientMap();
    const pieces = [recipe.id, recipe.name && recipe.name.zh, recipe.name && recipe.name.en]
      .concat(recipe.aliases || [])
      .concat((recipe.base || []).map(id => map[id] ? `${map[id].zh} ${map[id].en}` : id))
      .concat((recipe.ingredients || []).map(item => map[item.id] ? `${map[item.id].zh} ${map[item.id].en}` : item.id))
      .concat(recipe.taste || []);
    if (language) pieces.push(recipe.description && recipe.description[language]);
    return normalizeSearch(pieces.filter(Boolean).join(" "));
  }

  window.DD_CALC = { effectiveCapacity, estimateAmounts, estimateAbv, estimateCost, isOwned, requiredIngredientIds, matchRecipe, normalizeSearch, searchableText };
}());

File: android/app/src/main/java/com/zalandunk/dawnsdew/data/Calculators.kt
package com.zalandunk.dawnsdew.data

import java.text.Normalizer
import java.util.Locale

data class EstimatedIngredient(val ingredient: RecipeIngredient, val estimatedMl: Double)

data class RecipeEstimate(
    val capacity: Double,
    val totalMl: Double,
    val ingredients: List<EstimatedIngredient>,
    val abv: Double,
    val cost: Double,
    val costCoverage: Double,
    val hasCost: Boolean,
    val exceedsCapacity: Boolean
)

data class RecipeMatch(
    val required: List<String>,
    val missing: List<String>,
    val matched: Int
) {
    val ratio: Double get() = if (required.isEmpty()) 1.0 else matched.toDouble() / required.size
    val status: String get() = when (missing.size) {
        0 -> "makeable"
        1 -> "almost"
        else -> "missing"
    }
}

object RecipeCalculator {
    fun estimate(
        recipe: Recipe,
        settings: AppSettings,
        pantry: Map<String, PantryEntry>,
        ingredientMap: Map<String, Ingredient>
    ): RecipeEstimate {
        val capacity = if (recipe.ice == "full") settings.icedLiquidCapacity else recipe.glassCapacity.takeIf { it > 0 }
            ?: settings.glassCapacity
        var total = 0.0
        val amounts = recipe.ingredients.map { item ->
            val volume = when {
                item.unit == "ml" -> (item.amount ?: 0.0).coerceAtLeast(0.0)
                item.unit == "fill" || item.fillTo != null ->
                    (capacity * (item.fillTo ?: 0.0).coerceIn(0.0, 1.0) - total).coerceAtLeast(0.0)
                item.unit == "top" || item.topUp -> (capacity - total).coerceAtLeast(0.0)
                else -> 0.0
            }
            total += volume
            EstimatedIngredient(item, volume)
        }
        var pureAlcohol = 0.0
        var cost = 0.0
        var pricedVolume = 0.0
        var neededVolume = 0.0
        amounts.forEach { estimated ->
            val entry = pantry[estimated.ingredient.id]
            val catalogAbv = ingredientMap[estimated.ingredient.id]?.abv ?: 0.0
            val abv = entry?.abv?.toDoubleOrNull() ?: catalogAbv
            pureAlcohol += estimated.estimatedMl * abv / 100.0
            if (estimated.estimatedMl > 0.0) {
                neededVolume += estimated.estimatedMl
                val price = entry?.price?.toDoubleOrNull()
                val packSize = entry?.packSize?.toDoubleOrNull()
                if (price != null && price >= 0.0 && packSize != null && packSize > 0.0) {
                    cost += price / packSize * estimated.estimatedMl
                    pricedVolume += estimated.estimatedMl
                }
            }
        }
        return RecipeEstimate(
            capacity = capacity,
            totalMl = total,
            ingredients = amounts,
            abv = if (total > 0.0) pureAlcohol / total * 100.0 else 0.0,
            cost = cost,
            costCoverage = if (neededVolume > 0.0) pricedVolume / neededVolume else 0.0,
            hasCost = pricedVolume > 0.0,
            exceedsCapacity = total > capacity + 0.01
        )
    }

    fun isOwned(entry: PantryEntry?): Boolean {
        if (entry?.owned != true) return false
        return entry.stock.isBlank() || (entry.stock.toDoubleOrNull() ?: 0.0) > 0.0
    }

    fun match(recipe: Recipe, pantry: Map<String, PantryEntry>): RecipeMatch {
        val required = recipe.ingredients
            .filter { !it.optional && it.id != "ice" }
            .map { it.id }
            .distinct()
        val missing = required.filter { !isOwned(pantry[it]) }
        return RecipeMatch(required, missing, required.size - missing.size)
    }

    fun searchableText(recipe: Recipe, ingredientMap: Map<String, Ingredient>): String = normalizeSearch(
        buildList {
            add(recipe.id)
            add(recipe.name.zh)
            add(recipe.name.en)
            addAll(recipe.aliases)
            addAll(recipe.taste)
            recipe.base.forEach { id -> ingredientMap[id]?.let { add("${it.name.zh} ${it.name.en}") } ?: add(id) }
            recipe.ingredients.forEach { item -> ingredientMap[item.id]?.let { add("${it.name.zh} ${it.name.en}") } ?: add(item.id) }
            add(recipe.description.zh)
            add(recipe.description.en)
        }.joinToString(" ")
    )

    fun normalizeSearch(value: String): String = Normalizer.normalize(value, Normalizer.Form.NFD)
        .replace(Regex("\\p{Mn}+"), "")
        .lowercase(Locale.ROOT)
        .replace(Regex("[’'·()（）&+]"), " ")
        .replace(Regex("\\s+"), " ")
        .trim()
}
